src/surrogate_heat_driver.cpp

```diff
@@ -87,40 +180,244 @@ void SurrogateHeatDriver::generate_arrays()
+
+  // determine the power deposition in each channel; the target applications will always
+  // be steady-state or pseudo-steady-state cases with no axial conduction such that
+  // the power deposition in each channel is independent of a convective heat transfer
+  // coefficient and only depends on the rod power at that axial elevation. The channel
+  // powers are indexed by channel ID, axial ID
+  xt::xtensor<double, 2> channel_powers = xt::zeros<double>({n_channels_, n_axial_});
+  for (int i = 0; i < n_channels_; ++i) {
+    for (int j = 0; j < n_axial_; ++j) {
+      for (const auto& rod : channels_[i].rod_ids_)
+        channel_powers(i, j) += 0.25 * rod_axial_node_power(rod, j);
+    }
+  }
+
+  // initial guesses for the fluid solution are uniform temperature (set to the inlet
+  // temperature) and uniform pressure  (set to the outlet pressure). These solution
+  // fields are defined on channel axial faces
+  xt::xtensor<double, 2> h = xt::zeros<double>({n_channels_, n_axial_ + 1});
+  xt::xtensor<double, 2> p = xt::zeros<double>({n_channels_, n_axial_ + 1});
+  std::fill(h.begin(), h.end(), iapws::h1(pressure_bc_, inlet_temperature_));
+  std::fill(p.begin(), p.end(), pressure_bc_);
+
+  // for certain verbosity settings, we will need to save the velocity solutions
+  xt::xtensor<double, 2> u = xt::zeros<double>({n_channels_, n_axial_ + 1});
+
+  bool converged = false;
+  for (gsl::index iter = 0; iter < max_subchannel_its_; ++iter) {
+    // save the previous solution
+    xt::xtensor<double, 2> h_old = h;
+    xt::xtensor<double, 2> p_old = p;
+
+    // solve each channel independently
+    for (gsl::index chan = 0; chan < n_channels_; ++chan) {
+      const auto& c = channels_[chan];
+
+      // solve for enthalpy by simple energy balance q = mdot * dh by marching from inlet
+      h(chan, 0) = iapws::h1(p(chan, 0), inlet_temperature_);
+      for (gsl::index axial = 0; axial < n_axial_; ++axial)
+        h(chan, axial + 1) = h(chan, axial) + channel_powers(chan, axial) / channel_flowrates_(chan);
+
+      // solve for pressure using one-sided finite difference approximation by marching
+      // from outlet and solving the axial momentum equation.
+      p(chan, n_axial_) = pressure_bc_;
+      for (gsl::index axial = n_axial_; axial > 0; axial--) {
+        // nu1 returns specific volume in units of m^3/kg, convert to g/cm^3
+        double rho_high = 1.0e-3 / iapws::nu1(p(chan, axial), iapws::T_from_p_h(p(chan, axial), h(chan, axial)));
+        double rho_low = 1.0e-3 / iapws::nu1(p(chan, axial - 1), iapws::T_from_p_h(p(chan, axial - 1), h(chan, axial - 1)));
+
+        u(chan, axial) = channel_flowrates_(chan) / (rho_high * c.area_);
+        u(chan, axial - 1) = channel_flowrates_(chan) / (rho_low * c.area_);
+
+        p[axial - 1] = p[axial] + channel_flowrates_(chan) / c.area_ * (u(chan, axial) - u(chan, axial - 1)) +
+          g_ * (z_(axial) - z_(axial - 1)) * rho_low;
```

Comment:
Do the units really work out right on this? In the SMRT subchannel solver, there's a 1e-6 factor on the second term.

Reply:
nope, missed that one too. Thanks for the thorough review!